File: tools/ios-crawler/src/ocr.js
/**
 * On-device text recognition, compiled on demand from ocr.m.
 *
 * The same trick as frames.js: clang ships with the Command Line Tools and
 * Vision ships with macOS, so the crawler can read the text on a screenshot
 * without an API key, a network call, or anything to install.
 */

import { existsSync, mkdirSync, statSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';
import { has, run } from './exec.js';
import { log } from './log.js';

const HERE = dirname(fileURLToPath(import.meta.url));
const SOURCE = join(HERE, 'ocr.m');
const BIN_DIR = join(HERE, '..', '.bin');
const BINARY = join(BIN_DIR, 'motvin-ocr');

const FRAMEWORKS = ['Foundation', 'Vision', 'AppKit', 'CoreGraphics'];

export async function buildOcr() {
  if (existsSync(BINARY) && statSync(BINARY).mtimeMs >= statSync(SOURCE).mtimeMs) {
    return BINARY;
  }
  if (!(await has('clang'))) {
    throw new Error('clang is not available, so on-device text recognition cannot be built. Install the Command Line Tools: xcode-select --install');
  }

  mkdirSync(BIN_DIR, { recursive: true });
  log.detail('building the text reader (once)…');

  const result = await run(
    'clang',
    [
      '-fobjc-arc',
      '-O2',
      '-Wno-deprecated-declarations',
      ...FRAMEWORKS.flatMap((framework) => ['-framework', framework]),
      '-o',
      BINARY,
      SOURCE,
    ],
    { timeout: 120_000 },
  );
  if (result.failed) {
    throw new Error(`could not build the text reader: ${result.stderr.trim().split('\n').slice(0, 3).join(' ')}`);
  }
  return BINARY;
}

/**
 * @typedef {{text: string, x: number, y: number, w: number, h: number, confidence: number}} TextLine
 * y is measured from the top, all values normalised 0–1.
 */

/**
 * Reads the text on one screenshot.
 * @returns {Promise<TextLine[]>}
 */
export async function readText(imagePath) {
  const binary = await buildOcr();
  const result = await run(binary, [imagePath], { timeout: 120_000 });
  if (result.failed) {
    throw new Error(result.stderr.trim().split('\n')[0] || `text reader exited ${result.code}`);
  }

  const lines = [];
  for (const raw of result.stdout.split('\n')) {
    const trimmed = raw.trim();
    if (!trimmed.startsWith('{')) continue;
    try {
      const parsed = JSON.parse(trimmed);
      lines.push({
        text: String(parsed.t ?? '').trim(),
        x: Number(parsed.x) || 0,
        y: Number(parsed.y) || 0,
        w: Number(parsed.w) || 0,
        h: Number(parsed.h) || 0,
        confidence: Number(parsed.c) || 0,
      });
    } catch {
      // A malformed line is skipped rather than failing the screen.
    }
  }
  return lines.filter((line) => line.text.length > 0).sort((a, b) => a.y - b.y);
}

File: tools/ios-crawler/src/ocr.m
/*
 * motvin-ocr — read the text on a screenshot using only what macOS already has.
 *
 *   motvin-ocr <image>
 *
 * Emits one JSON object per line:
 *   {"t":"Continue","x":0.12,"y":0.78,"w":0.76,"h":0.03}
 *
 * Coordinates are normalised 0–1 with y measured from the TOP, which is how
 * everyone reasons about a screen. Vision reports from the bottom, so it is
 * flipped here rather than in three places downstream.
 *
 * This exists so the crawler can classify screens without an API key. Apple's
 * Vision text recogniser runs on-device, costs nothing, and needs no network —
 * which makes rule-based classification possible offline. Built on demand by
 * src/ocr.js and cached in .bin/.
 */

#import <AppKit/AppKit.h>
#import <Foundation/Foundation.h>
#import <Vision/Vision.h>

/** JSON string escaping — enough for the text Vision returns. */
static NSString *Escape(NSString *text) {
  NSMutableString *out = [NSMutableString stringWithCapacity:text.length + 8];
  for (NSUInteger i = 0; i < text.length; i++) {
    unichar c = [text characterAtIndex:i];
    switch (c) {
      case '"': [out appendString:@"\\\""]; break;
      case '\\': [out appendString:@"\\\\"]; break;
      case '\n': [out appendString:@"\\n"]; break;
      case '\r': [out appendString:@"\\r"]; break;
      case '\t': [out appendString:@"\\t"]; break;
      default:
        if (c < 0x20) {
          [out appendFormat:@"\\u%04x", c];
        } else {
          [out appendFormat:@"%C", c];
        }
    }
  }
  return out;
}

int main(int argc, const char **argv) {
  @autoreleasepool {
    if (argc < 2) {
      fprintf(stderr, "usage: motvin-ocr <image>\n");
      return 2;
    }

    NSString *path = [NSString stringWithUTF8String:argv[1]];
    NSImage *image = [[NSImage alloc] initWithContentsOfFile:path];
    if (!image) {
      fprintf(stderr, "could not read %s\n", argv[1]);
      return 1;
    }

    CGImageRef cgImage = [image CGImageForProposedRect:NULL context:nil hints:nil];
    if (!cgImage) {
      fprintf(stderr, "could not decode %s\n", argv[1]);
      return 1;
    }

    VNImageRequestHandler *handler = [[VNImageRequestHandler alloc] initWithCGImage:cgImage options:@{}];
    VNRecognizeTextRequest *request = [[VNRecognizeTextRequest alloc] init];
    // Accurate rather than fast: UI type is small and often low-contrast, and
    // one screenshot's worth of text is quick either way.
    request.recognitionLevel = VNRequestTextRecognitionLevelAccurate;
    request.usesLanguageCorrection = NO;  // App names and labels are not prose.

    NSError *error = nil;
    if (![handler performRequests:@[ request ] error:&error] || error) {
      fprintf(stderr, "vision failed: %s\n", error ? error.localizedDescription.UTF8String : "unknown");
      return 1;
    }

    for (VNRecognizedTextObservation *observation in request.results) {
      VNRecognizedText *best = [[observation topCandidates:1] firstObject];
      if (!best || best.string.length == 0) continue;
      CGRect box = observation.boundingBox;
      double yFromTop = 1.0 - (box.origin.y + box.size.height);
      printf("{\"t\":\"%s\",\"x\":%.4f,\"y\":%.4f,\"w\":%.4f,\"h\":%.4f,\"c\":%.3f}\n",
             Escape(best.string).UTF8String, box.origin.x, yFromTop, box.size.width, box.size.height,
             best.confidence);
    }

    return 0;
  }
}
